{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Callysto.ca Banner](https://github.com/callysto/curriculum-notebooks/blob/master/callysto-notebook-banner-top.jpg?raw=true)\n",
    "\n",
    "<a href=\"https://hub.callysto.ca/jupyter/hub/user-redirect/git-pull?repo=https%3A%2F%2Fgithub.com%2Fcallysto%2Fhackathon&branch=master&subPath=ColonizingMars/ChallengeTemplates/challenge-option-1-should-we-colonize-Mars.ipynb&depth=1\" target=\"_parent\"><img src=\"https://raw.githubusercontent.com/callysto/curriculum-notebooks/master/open-in-callysto-button.svg?sanitize=true\" width=\"123\" height=\"24\" alt=\"Open in Callysto\"/></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove the '#' below and run this line if you are having an error that says \"no module named Bio\"\n",
    "#!pip install biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# üöÄ Spacer's Challenge : Using Data and Biology to Decide What We Should Grow On Mars\n",
    "\n",
    "\n",
    "## üìå Mission Statement\n",
    "\n",
    "You are a top scientist sent from Earth tasked with the most important mission of all: ensuring our survival on other planets. Before sending you off to space, your team gave you a variety of crop seeds, however there's a problem! The bag containing the crop seeds split open and spilled everywhere during your space travel! You picked up the ones you could find, but you now have no idea what it grows or what is special about them.\n",
    "\n",
    "Now on Mars, it is your job to [decide which crops to grow](https://www.sciencealert.com/growing-potatoes-on-mars-might-actually-work-hints-a-new-experiment). We will be using some basic bioinformatics concepts to try and find this out. To help you with this, you will work with a special robot-helper who will help write code for you, named ***Crop-3PO***.\n",
    "![](http://biowyse.eu/wp-content/uploads/2016/05/agrospace_01.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "**ü§ñ Crop-3PO:** Welcome! The people of Mars are counting on you! In order to use the special command, we need to import a special library by typing `import pandas`. We'll also need a few other libraries. \n",
    "\n",
    "**ü§ñ Crop-3PO:** I wrote it down in the code cell below, however because I'm just a robot I can't execute the code! I'll need your help to do that. The easiest way to do this is by clicking `cell -> Run All`. If you don't see anything or think something's broken, you can click `Run`in the toolbar above while you have the code cell selected.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# import libraries\n",
    "import pandas \n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# import data\n",
    "cropdata = pandas.read_csv(\"./data/cropdatafromearth.csv\")\n",
    "traitdata = pandas.read_csv(\"./data/isaaa_croptraits.csv\")\n",
    "approvaldata = pandas.read_csv(\"./data/foodapprovals.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## üåΩ Before we get started, what even is a GMO? \n",
    "\n",
    "*According to [gmotesting.com](http://www.gmotesting.com/GMOs/What-is-a-GMO):*\n",
    "\n",
    "> **GMO** is an acronym for genetically modified organism and is commonly used to refer to genetically altered crops that are grown in many areas around the globe.  Genetic engineering provides the ability to confer desired traits on plants such as herbicide tolerance and/or virus or insect resistance.  \n",
    "\n",
    "> The process of genetic modification involves splicing, or cutting, genes from one organism, such as a bacterium, virus, or animal, and inserting them into a recipient organism, such as a plant, so that the recipient is now able to express new traits provided by the donor genes.  The genetic material (commonly called a transgene) is inserted into the nucleus of a plant cell where it integrates into the plant DNA.  If integration of the DNA is successful, the plant cell, now described as a transgenic cell, divides and grows into a genetically modified (or transgenic) plant.  The genetic modification is permanent and will be passed on to the seeds of the transgenic (GMO) plant."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **ü§ñ Crop-3PO:** GMOs have grown in popularity internationally over the last 30 years as  our scientific methods become more reliable. Take a look at some of these charts that show which countries started adopting GMOs earlier than others. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# Display a histogram of the number of approved GMOS by country\n",
    "fig = px.bar(approvaldata, x=\"Country\", y = '1992‚Äì2003', title=\"Number of Approved GMOs between 1992-2003\")\n",
    "fig.update_yaxes(range=[0, 400])\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# Display a histogram of the number of approved GMOS by country\n",
    "fig = px.bar(approvaldata, x=\"Country\", y = '2004‚Äì2014', title=\"Number of Approved GMOs between 2004-2014\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### üìö Look at the trends of approved GMOs by countries over the last three decades. What differences do you see? List some reasons why this might be.\n",
    "\n",
    "\n",
    "Fill in the boxes marked with ‚úèÔ∏è by double clicking on it and filling in your answer. Remember to save your answers by clicking `File > Save and Checkpoint`!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "‚úèÔ∏è\n",
    "\n",
    "‚úèÔ∏è\n",
    "\n",
    "‚úèÔ∏è"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## üåΩ How are GMOs made?\n",
    "\n",
    "![](https://i1.wp.com/sitn.hms.harvard.edu/wp-content/uploads/2015/08/Untitled2.png?w=900)\n",
    "Image from: http://sitn.hms.harvard.edu/flash/2015/how-to-make-a-gmo/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## üåΩ  What are plant traits? How are traits selected for?\n",
    "\n",
    "Plant traits are desirable phenotypes (observable expression of a gene, it's functional use), and can be almost anything as long as they are useful. Century after century, farmers and botanists have selected plants that have desirable traits (sweet taste, size, etc) and cross bred them over time to produce more of that crop. With the use of modern technology however, we have successfully determined various genes that result in different plant traits. Some traits are not just the addition or enhancement of something, but can also be a suppression (for instead, delayed ripening). Instead of breeding over many generations, we can directly insert the gene and have more direct and reliable results.\n",
    "\n",
    "To learn more about the science of how this is actually done, read here: https://www.nationalgeographic.com/environment/future-of-food/food-technology-gene-editing/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# Display a histogram breaking down the GMO varieties in most popular crops\n",
    "fig = px.histogram(cropdata, x=\"Crop\", color=\"Variety\", title = \"Breakdown of GMO Varieties in Top 90% of Crops\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## üåΩ How are plant genes and proteins related to traits? \n",
    "\n",
    "\n",
    "Genes are actually only one part of the story. The way a gene becomes useful is by being translated into a protein in the cell. This protein then goes on to have different functions based on it's physical properties. \n",
    "\n",
    "*From [The University of Kentucky's College of Agriculture:](https://entomology.ca.uky.edu/ef130)* \n",
    "> To transform a plant into a GMO plant, the gene that produces a genetic trait of interest is identified and separated from the rest of the genetic material from a donor organism. Most organisms have thousands of genes, a single gene represents only a tiny fraction of the total genetic makeup of an organism. \n",
    "\n",
    ">A donor organism may be a bacterium, fungus or even another plant. In the case of Bt corn, the donor organism is a naturally occurring soil bacterium, Bacillus thuringiensis, and the gene of interest produces a protein that kills Lepidoptera larvae, in particular, European corn borer. This protein is called the Bt delta endotoxin. Growers use Bt corn as an alternative to spraying insecticides for control of European and southwestern corn borer. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# histogram of different genes associated with traits\n",
    "fig = px.histogram(traitdata, x=\"Trait\", title=\"Number of genes associated with traits\", \n",
    "                   labels={'Trait':'Phenotypic Trait'}).update_xaxes(categoryorder=\"total descending\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### üìö Look at the distribution of traits in the above two graphs. Which traits are the most popular? Why do you think that is?  \n",
    "\n",
    "\n",
    "Fill in the boxes marked with ‚úèÔ∏è by double clicking on it and filling in your answer. Remember to save your answers by clicking `File > Save and Checkpoint`!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "‚úèÔ∏è\n",
    "\n",
    "‚úèÔ∏è\n",
    "\n",
    "‚úèÔ∏è"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **ü§ñ Crop-3PO:** As you can see below, there are many genes associated with traits! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# View the ENTIRE data frame\n",
    "pandas.set_option('display.max_rows', None)\n",
    "traitdata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# üß¨ üíª  Bioinformatics Hands-on portion"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that you understand a little more of what GMO's are, it's time to try and find out what our seeds are. Luckily, you have a sequencing machine that can tell you the DNA sequence of a seed. Crop-3PO also has a list of genes, but only knows a few of their DNA sequences. (Note that these sequences are made up, and that actual DNA sequences are much longer!)\n",
    "\n",
    "You sort your seeds out into 10 different groups based on colour, size, and hardness. You believe these ones are the same. By the end you have a total of 10 groups with only a few seeds each. To avoid wasting seeds, you tell Crop-3PO to sequence only one from each group. You call these your samples."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **ü§ñ Crop-3PO:** I know lots of genes and sequences! Here is an example of what `cry1Ab` looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# Here, we can see a gene sequence associated with cry1AB_1\n",
    "genes_dict = SeqIO.to_dict(SeqIO.parse(\"./data/genes.fa\", \"fasta\"))\n",
    "print(genes_dict[\"cry1Ab_1\"].format(\"fasta\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **ü§ñ Crop-3PO:** Here are all the genes I know, some I only know part of the sequence for. If it looks similar, you can act like they are the same as the trait. For instance `tNOS` can be considered `nos` for our purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "#it will also be helpful for us to get a list of all the genes and samples we have\n",
    "geneslist = [] \n",
    "geneslist.append(genes_dict.keys())\n",
    "geneslist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **ü§ñ Crop-3PO:** Here is a list of all the samples I successfully sequenced!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# list of our total samples\n",
    "samples_dict = SeqIO.to_dict(SeqIO.parse(\"./data/samples.fa\", \"fasta\"))\n",
    "sampleslist = [] \n",
    "sampleslist.append(samples_dict.keys())\n",
    "sampleslist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **ü§ñ Crop-3PO:** Here's an example of what the sequence for Sample 1 looks like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# let's look at some of the information in our first sample\n",
    "print(samples_dict[\"sample1\"].format(\"fasta\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **ü§ñ Crop-3PO:** That's probably pretty difficult for a human to read. Luckily, as a machine I can very quickly match the sequences of our genes and tell you whether a sample is likely to have the gene or not. Let me try doing this, and you can see my output down below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# now we're going to match the genes to each sample\n",
    "x=[]\n",
    "y=[]\n",
    "for sample in samples_dict: \n",
    "    x.append(samples_dict[sample].name)\n",
    "    y.append(samples_dict[sample].seq)\n",
    "    temp2 = []\n",
    "    for gene in genes_dict:\n",
    "        temp1 = []\n",
    "        if genes_dict[gene].seq in samples_dict[sample].seq: \n",
    "            temp1.append(genes_dict[gene].name)\n",
    "            temp2.append(temp1)\n",
    "    print(\"üî¨\\033[1;36m\", samples_dict[sample].name, \"\\033[1;0m contains matches for \\033[1;32m\", temp2, \"\\033[1;0m\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  **ü§ñ Crop-3PO:** On a larger scale, if we had sequence matches for every single gene, we would likely see something like this:\n",
    "![](./images/genome.png)\n",
    "\n",
    "Image obtained from: https://www.nature.com/articles/s41598-019-51668-x/figures/3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **ü§ñ Crop-3PO:** Pretty cool trick huh? Now all that's left to do is try to figure out which ones are useful for us. You'll probably want to look at the genes in each sample and think about which traits they're associated with, and whether or not that is useful for us to have here on Mars. I've listed them down below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ‚ùóÔ∏è Run this cell!\n",
    "# trait data narrowed down to our genes\n",
    "traitdata.loc[traitdata['Gene'].isin(['cry1Ab','nos','cry1Ac','cry2Ab2','epsps (Ag)','prsv_cp','cel1','bar','pat'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### üìö Based on the information available to us, what traits does each sample have?\n",
    "\n",
    "Fill in the boxes marked with ‚úèÔ∏è by double clicking on it and filling in your answer. Remember to save your answers by clicking `File > Save and Checkpoint`!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "‚úèÔ∏è\n",
    "\n",
    "‚úèÔ∏è\n",
    "\n",
    "‚úèÔ∏è"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### üìö Out of all of our samples, which ones do you think would be most suitable to grow on Mars? Explain your reasoning. Keep in mind, what might be beneficial on Earth might not make much of a difference on Mars!\n",
    "\n",
    "Fill in the boxes marked with ‚úèÔ∏è by double clicking on it and filling in your answer. Remember to save your answers by clicking `File > Save and Checkpoint`!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "‚úèÔ∏è\n",
    "\n",
    "‚úèÔ∏è\n",
    "\n",
    "‚úèÔ∏è"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### üìö Assuming the samples are the same plant (e.g. all are Corn, etc), which two crops would you try to cross first to achieve a \"stacked\" trait? \n",
    "\n",
    "### A stacked trait is a trait that has multiple benefits.\n",
    "\n",
    "Fill in the boxes marked with ‚úèÔ∏è by double clicking on it and filling in your answer. Remember to save your answers by clicking `File > Save and Checkpoint`!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "‚úèÔ∏è\n",
    "\n",
    "‚úèÔ∏è\n",
    "\n",
    "‚úèÔ∏è"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## üìå References \n",
    "#### The data used in this notebook was obtained from the following sources:\n",
    " \n",
    "Rhodora R Aldemita, Ian Mari E Rea√±o, Renando O Solis & Randy A Hautea (2015) Trends in global approvals of biotech crops (1992‚Äì2014), GM Crops & Food, 6:3, 150-166, DOI: 10.1080/21645698.2015.1056972 \n",
    "\n",
    "Debode, F., Hulin, J., Charloteaux, B. et al. Detection and identification of transgenic events by next generation sequencing combined with enrichment technologies. Sci Rep 9, 15595 (2019). https://doi.org/10.1038/s41598-019-51668-x\n",
    "\n",
    "ISAAA's GM Approval Database. http://www.isaaa.org/gmapprovaldatabase/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[![Callysto.ca License](https://github.com/callysto/curriculum-notebooks/blob/master/callysto-notebook-banner-bottom.jpg?raw=true)](https://github.com/callysto/curriculum-notebooks/blob/master/LICENSE.md)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
